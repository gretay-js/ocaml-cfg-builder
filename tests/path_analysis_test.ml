open Core
open Test_graph

module Paths = struct
  module Path = struct
    type t = ID.t list [@@deriving sexp, equal, compare]
  end

  module Set = Set.Make(Path)

  type t = Set.t [@@deriving sexp, equal]

  let lub = Set.union
end

let print_solution solution =
  ID.Map.iter
    (fun id (_, path_out) ->
      print_s [%message (id: ID.t)];
      Paths.Set.iter path_out ~f:(fun path ->
        print_string "  ";
        print_s [%message (path : Paths.Path.t)]))
    solution

let path_f _ node s =
  s
  |> Paths.Set.filter ~f:(fun path -> List.count path ~f:(Int.equal node) < 2)
  |> Paths.Set.map ~f:(fun path -> node :: path)

let () =
  let
    module Problem = struct
      module S = Paths

      let init t id =
        ((if is_entry t id then Paths.Set.singleton [] else Paths.Set.empty), Paths.Set.empty)

      let f = path_f
    end
  in
  let module Solver = Make_forward_graph_solver(Problem) in
  print_solution (Solver.solve simple_graph)

let () =
  let
    module Problem = struct
      module S = Paths

      let init t id =
        ((if is_exit t id then Paths.Set.singleton [] else Paths.Set.empty), Paths.Set.empty)

      let f = path_f
    end
  in
  let module Solver = Make_backward_graph_solver(Problem) in
  print_solution (Solver.solve simple_graph)

