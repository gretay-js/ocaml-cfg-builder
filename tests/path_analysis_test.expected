(id 1)
  (path (1))
(id 2)
  (path (2 1))
  (path (2 7 6 4 3 2 1))
  (path (2 7 6 5 3 2 1))
(id 3)
  (path (3 2 1))
  (path (3 2 7 6 4 3 2 1))
  (path (3 2 7 6 5 3 2 1))
(id 4)
  (path (4 3 2 1))
  (path (4 3 2 7 6 4 3 2 1))
  (path (4 3 2 7 6 5 3 2 1))
(id 5)
  (path (5 3 2 1))
  (path (5 3 2 7 6 4 3 2 1))
  (path (5 3 2 7 6 5 3 2 1))
(id 6)
  (path (6 4 3 2 1))
  (path (6 4 3 2 7 6 4 3 2 1))
  (path (6 4 3 2 7 6 5 3 2 1))
  (path (6 5 3 2 1))
  (path (6 5 3 2 7 6 4 3 2 1))
  (path (6 5 3 2 7 6 5 3 2 1))
(id 7)
  (path (7 6 4 3 2 1))
  (path (7 6 4 3 2 7 6 4 3 2 1))
  (path (7 6 4 3 2 7 6 5 3 2 1))
  (path (7 6 5 3 2 1))
  (path (7 6 5 3 2 7 6 4 3 2 1))
  (path (7 6 5 3 2 7 6 5 3 2 1))
(id 8)
  (path (8 7 6 4 3 2 1))
  (path (8 7 6 4 3 2 7 6 4 3 2 1))
  (path (8 7 6 4 3 2 7 6 5 3 2 1))
  (path (8 7 6 5 3 2 1))
  (path (8 7 6 5 3 2 7 6 4 3 2 1))
  (path (8 7 6 5 3 2 7 6 5 3 2 1))
(id 1)
  (path (1 2 3 4 6 7 2 3 4 6 7 8))
  (path (1 2 3 4 6 7 2 3 5 6 7 8))
  (path (1 2 3 4 6 7 8))
  (path (1 2 3 5 6 7 2 3 4 6 7 8))
  (path (1 2 3 5 6 7 2 3 5 6 7 8))
  (path (1 2 3 5 6 7 8))
(id 2)
  (path (2 3 4 6 7 2 3 4 6 7 8))
  (path (2 3 4 6 7 2 3 5 6 7 8))
  (path (2 3 4 6 7 8))
  (path (2 3 5 6 7 2 3 4 6 7 8))
  (path (2 3 5 6 7 2 3 5 6 7 8))
  (path (2 3 5 6 7 8))
(id 3)
  (path (3 4 6 7 2 3 4 6 7 8))
  (path (3 4 6 7 2 3 5 6 7 8))
  (path (3 4 6 7 8))
  (path (3 5 6 7 2 3 4 6 7 8))
  (path (3 5 6 7 2 3 5 6 7 8))
  (path (3 5 6 7 8))
(id 4)
  (path (4 6 7 2 3 4 6 7 8))
  (path (4 6 7 2 3 5 6 7 8))
  (path (4 6 7 8))
(id 5)
  (path (5 6 7 2 3 4 6 7 8))
  (path (5 6 7 2 3 5 6 7 8))
  (path (5 6 7 8))
(id 6)
  (path (6 7 2 3 4 6 7 8))
  (path (6 7 2 3 5 6 7 8))
  (path (6 7 8))
(id 7)
  (path (7 2 3 4 6 7 8))
  (path (7 2 3 5 6 7 8))
  (path (7 8))
(id 8)
  (path (8))
